File: New/atom.py

from re import sub


class atom:
    
    def __init__(self, letter, subscript=-1, power=1):                                              # Initializes atom class with require variables and defaults
        self.letter = letter                                                                        # Letter should be a single character
        self.subscript = subscript                                                                  # subscript should be an int
        self.power = power                                                                          # Power should be an int
    

    # These are pretty self explanitory - I'm also pretty sure these are not needed because you can do class.variable to get the variable and class.variable = x to set the class variables
    # def getLetter(self):
    #     return self.letter

    # def getSubscript(self):
    #     return self.subscript

    # def getPower(self):
    #     return self.power
    
    # def setLetter(self, letter):
    #     self.letter = letter
    
    # def setSubscript(self, subscript):
    #     self.subscript = subscript

    # def setPower(self, power):
    #     self.power = power

    # def setAtom(self, letter, subscript, power):
    #     self.letter = letter
    #     self.subscript = subscript
    #     self.power = power
    
    # This function presumably mutiplies two atoms by adding the exponent. It assumes that the subscripts and letters are the same as it is called time "like" atom
    def timesLikeAtom(self, atom):
        # For some reason doing "return atom(args)" did not work. There are better solutions but this one worked for some reason
        temp = atom
        temp.__init__(self.letter, self.subscript, self.power + atom.power)
        return temp

    # Returns true if the atoms are "like"
    def like(self, atom):   
        return self.letter == atom.letter and self.subscript == atom.subscript

    # Compares if two atoms are equal
    def equals(self, atom):
        return self.letter == atom.letter and self.subscript == atom.subscript and self.power == atom.power

    # Compares two atoms too see if the atoms is less or equal to the inputed atom
    def lessThanOrEqual(self, atom):
        if self.letter < atom.letter:                                                               # First checks if the letter itself is bigger
            return True
        elif self.letter == atom.letter:                                                            # If the leters are equal it first checks the subscripts to see if one is less
            if self.subscript < atom.subscript:
                return True
            elif self.subscript == atom.subscript and self.power < atom.power:                      # After checking the subscipts, if they are equal then it checks the powers
                return True
        return self.equals(atom)                                                                    # Kinda like a catch all - the previous code checks for less than and this just confirms if they are equal or not
        # It would have made more sense to use less than or equal too "<=" or ">=" instead of using the semi catch all at the lend

x = atom("a", power=0)
y = atom("a")

print(x.lessThanOrEqual(y))

File: Old/Atom.java
//
// Source code recreated from a .class file by IntelliJ IDEA
// (powered by Fernflower decompiler)
//


public class Atom implements Comparable<Atom> {
    private char letter;
    private int subscript = -1;
    private int power = 1;

    public Atom() {
    }

    public Atom(char letter, int subscript, int power) {
        this.letter = letter;
        this.subscript = subscript;
        this.power = power;
    }

    public Atom(char letter) {
        this.letter = letter;
    }

    public char getLetter() {
        return this.letter;
    }

    public int getSubscript() {
        return this.subscript;
    }

    public int getPower() {
        return this.power;
    }

    public void setLetter(char letter) {
        this.letter = letter;
    }

    public void setSubscript(int subscript) {
        this.subscript = subscript;
    }

    public void setPower(int power) {
        this.power = power;
    }

    public void setAtom(char letter, int subscript, int power) {
        this.letter = letter;
        this.subscript = subscript;
        this.power = power;
    }

    public Atom timesLikeAtom(Atom atom) {
        return new Atom(this.getLetter(), this.subscript, this.getPower() + atom.getPower());
    }

    /** @deprecated */
    @Deprecated
    public boolean like(Atom atom) {
        return this.getLetter() == atom.getLetter() && this.getSubscript() == atom.getSubscript();
    }

    public boolean isLike(Atom atom) {
        return this.getLetter() == atom.getLetter() && this.getSubscript() == atom.getSubscript();
    }

    /** @deprecated */
    @Deprecated
    public boolean lessThanOrEqual(Atom atom) {
        if (this.getLetter() < atom.getLetter()) {
            return true;
        } else {
            if (this.getLetter() == atom.getLetter()) {
                if (this.getSubscript() < atom.getSubscript()) {
                    return true;
                }

                if (this.getSubscript() == atom.getSubscript() && this.getPower() < atom.getPower()) {
                    return true;
                }
            }

            return this.equals(atom);
        }
    }

    public boolean isLessThanOrEquals(Atom atom) {
        if (this.equals(atom)) {
            return true;
        } else if (this.getLetter() < atom.getLetter()) {
            return true;
        } else {
            if (this.getLetter() == atom.getLetter()) {
                if (this.getSubscript() < atom.getSubscript()) {
                    return true;
                }

                if (this.getSubscript() == atom.getSubscript() && this.getPower() < atom.getPower()) {
                    return true;
                }
            }

            return false;
        }
    }

    /** @deprecated */
    @Deprecated
    public boolean lessThan(Atom atom) {
        if (this.getLetter() < atom.getLetter()) {
            return true;
        } else {
            return this.getLetter() == atom.getLetter() && this.getSubscript() < atom.getSubscript();
        }
    }

    public boolean isLessThan(Atom atom) {
        if (this.getLetter() < atom.getLetter()) {
            return true;
        } else {
            return this.getLetter() == atom.getLetter() && this.getSubscript() < atom.getSubscript();
        }
    }

    /** @deprecated */
    @Deprecated
    public boolean identicalTo(Atom atom) {
        return this.getLetter() == atom.getLetter() && this.getSubscript() == atom.getSubscript() && this.getPower() == atom.getPower();
    }

    /** @deprecated */
    @Deprecated
    public void print() {
        if (this.subscript == -1) {
            if (this.power == 1) {
                System.out.print(this.letter);
            } else if (this.power != 0) {
                System.out.print(this.letter + "^" + this.power);
            }
        } else if (this.power == 1) {
            System.out.print(this.letter + "_" + this.subscript);
        } else if (this.power != 0) {
            System.out.print(this.letter + "_" + this.subscript + "^" + this.power);
        }

    }

    public boolean equals(Atom atom) {
        return this.getLetter() == atom.getLetter() && this.getSubscript() == atom.getSubscript() && this.getPower() == atom.getPower();
    }

    public String toString() {
        String string = "";
        if (this.subscript == -1) {
            if (this.power == 1) {
                string = String.valueOf(this.letter);
            } else if (this.power != 0) {
                string = this.letter + "^" + this.power;
            }
        } else if (this.power == 1) {
            string = this.letter + "_" + this.subscript;
        } else if (this.power != 0) {
            string = this.letter + "_" + this.subscript + "^" + this.power;
        }

        return string;
    }

    public int compareTo(Atom a) {
        if (this.equals(a)) {
            return 0;
        } else {
            return this.isLessThan(a) ? -1 : 1;
        }
    }
}
